### ISSUE TITLE
BrownCompleteness can't reload cached file

### ISSUE THREAD

**Issue by rhondamorgan** [dsavransky/EXOSIMS#27]
I performed a newclone of EXOSIMS yesterday.  It crashes when using BrownCompleteness.  
I intend to use BrownCompleteness until unit testing is performed on GarrettCompleteness.

Here is the screen output and error message: 

MissionSim: Seed is:  686326949
Imported SurveyEnsemble (prototype module) from EXOSIMS.Prototypes.SurveyEnsemble
Imported SimulatedUniverse (prototype module) from EXOSIMS.Prototypes.SimulatedUniverse
Imported TargetList (prototype module) from EXOSIMS.Prototypes.TargetList
Imported EXOCAT1 (specific module) from EXOSIMS.StarCatalog.EXOCAT1
Imported KasdinBraems (specific module) from EXOSIMS.OpticalSystem.KasdinBraems
Imported Stark (specific module) from EXOSIMS.ZodiacalLight.Stark
Imported PostProcessing (prototype module) from EXOSIMS.Prototypes.PostProcessing
Imported GalaxiesFaintStars (specific module) from EXOSIMS.BackgroundSources.GalaxiesFaintStars
Imported BrownCompleteness (specific module) from EXOSIMS.Completeness.BrownCompleteness
Imported EarthTwinHabZone1 (specific module) from EXOSIMS.PlanetPopulation.EarthTwinHabZone1
Imported PlanetPhysicalModel (prototype module) from EXOSIMS.Prototypes.PlanetPhysicalModel
Loading cached completeness file from "/proj/exep/rhonda/exosimsgit2017/EXOSIMS/EXOSIMS/Completeness/EarthTwinHabZone10cdf3ce3f4d97068fea6468ef03add60.comp".
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-6-af9af77f0f9d> in <module>()
----> 1 sim = EXOSIMS.MissionSim.MissionSim(scriptfile)

/proj/exep/rhonda/exosimsgit2017/EXOSIMS/EXOSIMS/MissionSim.pyc in __init__(self, scriptfile, **specs)
    110         self.modules = {}
    111         self.modules['SimulatedUniverse'] = get_module(specs['modules']\
--> 112                 ['SimulatedUniverse'],'SimulatedUniverse')(**specs)
    113         self.modules['Observatory'] = get_module(specs['modules']\
    114                 ['Observatory'],'Observatory')(**specs)

/proj/exep/rhonda/exosimsgit2017/EXOSIMS/EXOSIMS/Prototypes/SimulatedUniverse.pyc in __init__(self, **specs)
     90 
     91         # import TargetList class
---> 92         self.TargetList = get_module(specs['modules']['TargetList'],'TargetList')(**specs)
     93 
     94         # bring inherited class objects to top level of Simulated Universe

/proj/exep/rhonda/exosimsgit2017/EXOSIMS/EXOSIMS/Prototypes/TargetList.pyc in __init__(self, keepStarCatalog, minComp, **specs)
     93 
     94         # now populate and filter the list
---> 95         self.populate_target_list(**specs)
     96         self.filter_target_list(**specs)
     97 

/proj/exep/rhonda/exosimsgit2017/EXOSIMS/EXOSIMS/Prototypes/TargetList.pyc in populate_target_list(self, **specs)
    148         self.nan_filter()
    149         # populate completeness values
--> 150         self.comp0 = Comp.target_completeness(self)
    151         # populate minimum integration time values, for minimum dMag in detection mode
    152         mode = filter(lambda mode: mode['detectionMode'] == True, OS.observingModes)[0]

/proj/exep/rhonda/exosimsgit2017/EXOSIMS/EXOSIMS/Completeness/BrownCompleteness.pyc in target_completeness(self, TL)
    100         # path to 2D completeness pdf array for interpolation
    101         Cpath = os.path.join(self.classpath, self.filename+'.comp')
--> 102         Cpdf, xedges2, yedges2 = self.genC(Cpath, nplan, xedges, yedges, steps)
    103 
    104         EVPOCpdf = interpolate.RectBivariateSpline(xedges, yedges, Cpdf.T)

/proj/exep/rhonda/exosimsgit2017/EXOSIMS/EXOSIMS/Completeness/BrownCompleteness.pyc in genC(self, Cpath, nplan, xedges, yedges, steps)
    286         if os.path.exists(Cpath):
    287             print 'Loading cached completeness file from "%s".' % Cpath
--> 288             H = pickle.load(open(Cpath, 'rb'))
    289             print 'Completeness loaded from cache.'
    290             #h, xedges, yedges = self.hist(nplan, xedges, yedges)

ValueError: buffer size does not match array size

**Comment by dgarrett622** [comment]
I have not been able to reproduce this behavior. I suspect it has to do with pickle. Did BrownCompleteness generate the .comp file from your new clone or did you copy/paste from an existing installation? 

Your pickle installation may be newer than the one used to save the file if you copy/pasted it. If that is the case, this happens because pickle is not backward compatible. BrownCompleteness loads a numpy array saved using pickle. If portability is an issue, I could look into saving these arrays in a different way.

**Comment by rhondamorgan** [comment]
Hi Daniel,
This error occurred the first time I ran this new installation, so there was no .comp file to load.  Potentially the array size error is because the filename is empty?
Thank you,
Rhonda

**Comment by dgarrett622** [comment]
The line where the issue occurred is after BrownCompleteness checks for an existing .comp file. Are you sure there is no .comp file in the Completeness directory? 

EXOSIMS printed this to the screen from your output:
Loading cached completeness file from "/proj/exep/rhonda/exosimsgit2017/EXOSIMS/EXOSIMS/Completeness/EarthTwinHabZone10cdf3ce3f4d97068fea6468ef03add60.comp".

**Comment by rhondamorgan** [comment]
Hi Daniel,
I’m very sure.
I cloned that latest git version into new directory.
I attempted to run it.  I got an error because the des432s.bsp file was missing.  I copied it in.
I attempted to run, as you see on the screen.  genC did not run as I expected it to when no *.comp is present.  I got the error I reported.
Rhonda

**Comment by dgarrett622** [comment]
Okay, I still can't reproduce the issue you are seeing. I cloned the latest git version of EXOSIMS to a new directory and ran it. It generated the .comp file on the first run and then loaded the .comp file on subsequent runs. @ChrisDelaX or @dsavransky are you able to reproduce the issue Rhonda raised?

**Comment by rhondamorgan** [comment]
I’ll try to reproduce this when my current run of 1000 DRMs on an 8m telescope with ~950 targets finishes…probably in 10 hours (42 cores).
Rhonda

**Comment by dsavransky** [comment]
@rhondamorgan Can you please tell us the OS & python versions of the system you're running on.  Also, are you using pickle or cPickle?

**Comment by rhondamorgan** [comment]
Hi Daniel,
I tried to replicate the error, and this is what I learned.


  1.  I attempted to run using BrownCompleteness and received the previous error.
  2.  I deleted the *.comp file and the BrownCompleteness.pyc file.  I ran and did not receive any errors.

Since you also have not been able to replicate the error, I am going to surmise that something was corrupt in the .pyc file or the .comp file, and the error did not reside in the code.

Thank you for taking the time to work on this bug and to try to replicate the error.

Rhonda

From: "Morgan, Rhonda M (383A)" <rhonda.m.morgan@jpl.nasa.gov<mailto:rhonda.m.morgan@jpl.nasa.gov>>
Date: Friday, March 10, 2017 at 1:39 PM
To: dsavransky/EXOSIMS <reply@reply.github.com<mailto:reply@reply.github.com>>, dsavransky/EXOSIMS <EXOSIMS@noreply.github.com<mailto:EXOSIMS@noreply.github.com>>
Cc: "Morgan, Rhonda M (383A)" <rhonda.m.morgan@jpl.nasa.gov<mailto:rhonda.m.morgan@jpl.nasa.gov>>, Author <author@noreply.github.com<mailto:author@noreply.github.com>>
Subject: Re: [dsavransky/EXOSIMS] BrownCompleteness can't reload cached file (#27)

I’ll try to reproduce this when my current run of 1000 DRMs on an 8m telescope with ~950 targets finishes…probably in 10 hours (42 cores).
Rhonda
